src/mcmc/gibbs.jl

```diff
@@ -1,251 +1,528 @@
+        )
+    end
+    return num_conditioned > 0
 end
 
-# Use PG for a 'v2' variable, and use HMC for the 'v1' variable.
-# Note that v2 is discrete, so the PG sampler is more appropriate
-# than is HMC.
-alg = Gibbs(HMC(0.2, 3, :v1), PG(20, :v2))
-```
+function get_conditioned_gibbs(context::GibbsContext, vn::VarName)
+    return get_global_varinfo(context)[vn]
+end
+function get_conditioned_gibbs(context::GibbsContext, vns::AbstractArray{<:VarName})
+    return map(Base.Fix1(get_conditioned_gibbs, context), vns)
+end
 
-One can also pass the number of iterations for each Gibbs component using the following syntax:
-- `alg = Gibbs((HMC(0.2, 3, :v1), n_hmc), (PG(20, :v2), n_pg))`
-where `n_hmc` and `n_pg` are the number of HMC and PG iterations for each Gibbs iteration.
+function is_target_varname(context::GibbsContext, vn::VarName)
+    return Iterators.any(
+        Iterators.map(target -> subsumes(target, vn), context.target_varnames)
+    )
+end
 
-Tips:
-- `HMC` and `NUTS` are fast samplers and can throw off particle-based
-methods like Particle Gibbs. You can increase the effectiveness of particle sampling by including
-more particles in the particle sampler.
-"""
-struct Gibbs{space,N,A<:NTuple{N,TGIBBS},B<:NTuple{N,Int}} <: InferenceAlgorithm
-    algs::A   # component sampling algorithms
-    iterations::B
-    function Gibbs{space,N,A,B}(
-        algs::A, iterations::B
-    ) where {space,N,A<:NTuple{N,TGIBBS},B<:NTuple{N,Int}}
-        all(isgibbscomponent, algs) ||
-            error("all algorithms have to support Gibbs sampling")
-        return new{space,N,A,B}(algs, iterations)
+function is_target_varname(context::GibbsContext, vns::AbstractArray{<:VarName})
+    num_target = count(Iterators.map(Base.Fix1(is_target_varname, context), vns))
+    if (num_target != 0) && (num_target != length(vns))
+        error(
+            "Some but not all of the variables in `vns` are target variables. " *
+            "Having mixed targeting like this is not supported in GibbsContext.",
+        )
     end
+    return num_target > 0
 end
 
-function Gibbs(alg1::TGIBBS, algrest::Vararg{TGIBBS,N}) where {N}
-    algs = (alg1, algrest...)
-    iterations = ntuple(Returns(1), Val(N + 1))
-    # obtain space for sampling algorithms
-    space = Tuple(union(getspace.(algs)...))
-    return Gibbs{space,N + 1,typeof(algs),typeof(iterations)}(algs, iterations)
+# Tilde pipeline
+function DynamicPPL.tilde_assume(context::GibbsContext, right, vn, vi)
+    if is_target_varname(context, vn)
+        # Fall back to the default behavior.
+        return DynamicPPL.tilde_assume(DynamicPPL.childcontext(context), right, vn, vi)
+    elseif has_conditioned_gibbs(context, vn)
+        # Short-circuit the tilde assume if `vn` is present in `context`.
+        value = get_conditioned_gibbs(context, vn)
+        # TODO(mhauru) Is the call to logpdf correct if context.context is not
+        # DefaultContext?
+        return value, logpdf(right, value), vi
```

Comment:
I'm trying to think of a way to avoid inserting contexts in the middle of the context stack. Could we instead do something like
```julia
child = childcontext(context)
while child isa SamplingContext
    child =  childcontext(child)
end
return DynamicPPL.tilde_assume(
            rng, child, right, vn, vi
        )
```
? So drop the `sampler` and any `SamplingContext`, and then call the usual `tilde_assume` pipeline. I'm not sure how the `sampler` argument and `SamplingContext` interact, but might also be necessary to go through the whole context stack and remove `SamplingContexts` anywhere they appear, not just at the top of the stack.

Reply:
Oh I was thinking more of doing the following:

```julia
function make_conditional(
    model::DynamicPPL.Model, target_variables::AbstractVector{<:VarName}, varinfo
)
    # Insert the `GibbsContext` just before the leaf.
    # 1. Extract the `leafcontext` from `model` and wrap in `GibbsContext`.
    gibbs_context_inner = GibbsContext(target_variables, Ref(varinfo), leafcontext(model.context))
    # 2. Set the leaf context to be the `GibbsContext` wrapping `leafcontext(model.context)`.
    gibbs_context = DynamicPPL.setleafcontext(model.context, gibbs_context_inner)
    return DynamicPPL.contextualize(model, gibbs_context), gibbs_context
end
```

You could also accompany this with something like

```julia
function GibbsContext(target_variables, varinfo_ref, context)
    # Check that the provided context is fact a leaf.
    DynamicPPL.NodeTrait(context) isa IsLeaf && throw(ArgumentError("`GibbsContext` can only wrap a leaf context"))
    ...
end
```

in the constructor.

Reply:
Is there a reason why that's preferable to what I have implemented, which is this:
```julia
    elseif has_conditioned_gibbs(context, vn)
        child = DynamicPPL.childcontext(context)
        if child isa SamplingContext
            # TODO(mhauru) Would it ever be valid to have a SamplingContext as the child?
            # We could just raise a warning, or optionally go down the stack of contexts
            # skipping all SamplingContexts. The erroring is being conservative.
            error("GibbsContext has a SamplingContext as its child.")
        end
        value, lp, _ = DynamicPPL.tilde_assume(
            child, right, vn, get_global_varinfo(context)
        )
        value, lp, vi
    else
```
What I like about this solution, is that it follows more the usual pattern of "do something and call the tilde pipeline on the childcontext". It also avoids inserting contexts in between other contexts, which feels less risky. For instance, if `GibbsContext` goes through the context stack and inserts it right above the leaf, then if any other context also needs to directly above the leaf for whatever reason, that's going to cause trouble, possibly the silent unexpected behavior sort of trouble. The above solution rather opts for the "do the simple, natural thing and error loudly if it might go wrong".

Reply:
With your suggestion we'll have to do remove _all_ `SamplingContext`s in the context stack, not just the `child` as the `SamplingContext` might be further down the stack.

> then if any other context also needs to directly above the leaf for whatever reason, that's going to cause trouble, possibly the silent unexpected behavior sort of trouble.

Making the constructor of `GibbsContext` error if we give it a context whichi sn't a leaf would address this, no?

Assuming we address the above by issue (i.e. look at entire stack rather than just `child`), you'll have to put that check in every tilde-statement. I was thinking it would be better to just deal with it _once_ in the "beginning" and then you're good 🤷 I'm also a bit worried about passing the `global_varinfo` into a pipeline which expects the "local" varinfo (I can't think of any cases right now that would cause issues here, _but_ seems non-ideal to me).

Reply:
> With your suggestion we'll have to do remove all SamplingContexts in the context stack, not just the child as the SamplingContext might be further down the stack.

Yeah, that's true. Or rather I might error if any `SamplingContext`s are found in the stack below. There's no valid reason to have one, right?

> Making the constructor of GibbsContext error if we give it a context whichi sn't a leaf would address this, no?

You still couldn't combine `GibbsContext` with any other context that depends on particularly positioning in the context stack, such as another context that would also always need to be at the very bottom, right above the leaf.

> Assuming we address the above by issue (i.e. look at entire stack rather than just child), you'll have to put that check in every tilde-statement.

I could (and should) move the check to constructor time, when we insert the GibbsContext into the stack. Or are you worried that something might mess with the context stack after the GibbsContext has been created? Not sure if this could happen, given the `GibbsContext ` is created within the call to `step`.

> I'm also a bit worried about passing the global_varinfo into a pipeline which expects the "local" varinfo

On the other hand, I like the fact that in my solution we call `tilde_assume` with some varinfo and the child context. Manually putting in a call to `logpdf` feels a bit like its duplicating code from much lower in down in DPPL, but at a much higher level in the sampler implementation. There's a certain neatness to saying "we call `tilde_assume` with the child context, we only vary whether we do the call with the `sampler` and the local varinfo, or without the sampler and with the global varinfo".

Reply:
> Manually putting in a call to logpdf feels a bit like its duplicating code from much lower in down in DPPL, but at a much higher level in the sampler implementation. 

So if we do the "parent to a leaf" approach, then you could still just defer evaluation to the child context:) In fact, that was part of the reason why I wanted to do this.

Reply:
> Or rather I might error if any SamplingContexts are found in the stack below. There's no valid reason to have one, right?

Hmm, I _think_ so, but can't say 100%

Reply:
Seems this boils down to which one is more likely, that we'll need to have a SamplingContext under a GibbsContext, or that another context is going to be upset by GibbsContext insisting on inserting itself right above the leaf.

I'm parking this for a moment though, because I currently have a more pressing issue where I seem to have lost type stability, and I don't know why...

Reply:
I went with your solution in the end, because it seems more fail-safe: You don't need to watch for modifications to the whole context stack, just the childcontext. We can revisit if this ever becomes prohibitively strict.

Fun fact: Using the wrong context here, like we were originally doing by essentially hard-coding DefaultContext, was masking a bug in `setparams_varinfo!!` for ESS. Unmasking the bug caused ESS to go into an infinite loop. You didn't need to know this, but I needed to tell someone, for therapeutic reasons, because it took me a day to figure this out.

Reply:
I think this should now be fixed, but leaving this thread open until a) you've had a look at the new implementation, b) I've added a test that would have caught this.